File: platform/ios/demo/Examples/ObjectiveC/PointConversionExample.m
#import "PointConversionExample.h"
@import Mapbox;

NSString *const MBXExamplePointConversion = @"PointConversionExample";

@interface PointConversionExample ()
@property (nonatomic) MGLMapView *mapView;
@end

@implementation PointConversionExample

- (void)viewDidLoad {
    [super viewDidLoad];

    self.mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds];
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;

    [self.view addSubview:self.mapView];

    // Double tapping zooms the map, so ensure that can still happen.
    UITapGestureRecognizer *doubleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:nil];
    doubleTap.numberOfTapsRequired = 2;
    [self.mapView addGestureRecognizer:doubleTap];

    // Delay single tap recognition until it is clearly not a double.
    UITapGestureRecognizer *singleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleSingleTap:)];
    [singleTap requireGestureRecognizerToFail:doubleTap];
    [self.mapView addGestureRecognizer:singleTap];

    // Convert `mapView.centerCoordinate` (CLLocationCoordinate2D)
    // to screen location (CGPoint).
    CGPoint centerScreenPoint = [self.mapView convertCoordinate:self.mapView.centerCoordinate
                                                  toPointToView:self.mapView];

    NSLog(@"Screen center: %@ = %@",
          NSStringFromCGPoint(centerScreenPoint),
          NSStringFromCGPoint(self.mapView.center));
}

- (void)handleSingleTap:(UITapGestureRecognizer *)tap {
    // Convert tap location (CGPoint)
    // to geographic coordinates (CLLocationCoordinate2D).
    CLLocationCoordinate2D location = [self.mapView convertPoint:[tap locationInView:self.mapView]
                                            toCoordinateFromView:self.mapView];

    NSLog(@"You tapped at: %.5f, %.5f", location.latitude, location.longitude);

    // Create an array of coordinates for our polyline.
    CLLocationCoordinate2D coordinates[] = {
        self.mapView.centerCoordinate,
        location
    };
    NSUInteger numberOfCoordinates = sizeof(coordinates) / sizeof(CLLocationCoordinate2D);

    // Remove existing polyline from the map, (re)add polyline with coordinates.
    if (self.mapView.annotations.count) {
        [self.mapView removeAnnotations:self.mapView.annotations];
    }
    MGLPolyline *polyline = [MGLPolyline polylineWithCoordinates:coordinates
                                                           count:numberOfCoordinates];
    [self.mapView addAnnotation:polyline];
}

@end

File: platform/ios/demo/Examples/Swift/PointConversionExample.swift
import Mapbox

@objc(PointConversionExample_Swift)

class PointConversionExample_Swift: UIViewController, MGLMapViewDelegate {
    var mapView: MGLMapView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        
        view.addSubview(mapView)
        
        // double tapping zooms the map, so ensure that can still happen
        let doubleTap = UITapGestureRecognizer(target: self, action: nil)
        doubleTap.numberOfTapsRequired = 2
        mapView.addGestureRecognizer(doubleTap)
        
        // delay single tap recognition until it is clearly not a double
        let singleTap = UITapGestureRecognizer(target: self, action: #selector(handleSingleTap))
        singleTap.require(toFail: doubleTap)
        mapView.addGestureRecognizer(singleTap)
        
        // convert `mapView.centerCoordinate` (CLLocationCoordinate2D)
        // to screen location (CGPoint)
        let centerScreenPoint: CGPoint = mapView.convert(mapView.centerCoordinate, toPointTo: mapView)
        print("Screen center: \(centerScreenPoint) = \(mapView.center)")
    }
    
    func handleSingleTap(tap: UITapGestureRecognizer) {
        // convert tap location (CGPoint)
        // to geographic coordinates (CLLocationCoordinate2D)
        let location: CLLocationCoordinate2D = mapView.convert(tap.location(in: mapView), toCoordinateFrom: mapView)
        print("You tapped at: \(location.latitude), \(location.longitude)")
        
        // create an array of coordinates for our polyline
        var coordinates: [CLLocationCoordinate2D] = [mapView.centerCoordinate, location]
        
        // remove existing polyline from the map, (re)add polyline with coordinates
        if (mapView.annotations?.count != nil) {
            mapView.removeAnnotations(mapView.annotations!)
        }

        let polyline = MGLPolyline(coordinates: &coordinates, count: UInt(coordinates.count))

        mapView.addAnnotation(polyline)
    }
}
